projects/ngx-drag-to-select/src/lib/select-item.spec.ts

```diff
@@ -0,0 +1,60 @@
+import { Component, ViewChild } from '@angular/core';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { DragToSelectModule } from './drag-to-select.module';
+import { SelectItemDirective } from './select-item.directive';
+
+const ITEM_VALUE = 1;
+
+@Component({
+  template: `
+    <dts-select-container>
+      <span [dtsSelectItem]="${ITEM_VALUE}"
+        #selectItem="dtsSelectItem"
+        (select)="itemSelected($event)"
+        (deselect)="itemDeselected($event)">Select me!</span>
+    </dts-select-container>
+  `
+})
+class TestComponent {
+  @ViewChild('selectItem') selectItem: SelectItemDirective;
+
+  itemSelected(value: any) {}
+  itemDeselected(value: any) {}
+}
+
+describe('SelectItemDirective', () => {
+  let fixture: ComponentFixture<TestComponent>;
+  let selectItem: SelectItemDirective;
+  let testComponent: TestComponent;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [TestComponent],
+      imports: [DragToSelectModule.forRoot()]
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    window.getSelection = jest.fn().mockReturnValue({});
```

Comment:
Hm, good question. I think it makes more sense to have them on the item directive because those are the ones being selected. Besides, there is already a `select` output on the container that emits when a new selection was made. The payload is the current list of select items. So it would be a bit ambiguous. We could solve this by renaming the events but I personally thing it's better to have them on the item directive.

What do you think? Why would you think it'd be better to have the container emit those events?

Reply:
Because that's what I originally thought 😅 mainly because it's possible to select/deselect via the container in #50 .
I agree it makes more sense to add it to the directive after seeing it.

Reply:
Hm true, you use the container to select items but the output is emitted on the directive... a little bit inconsistant. So do we go with this or should I move the events onto the container?

Reply:
If we would put it on the container which naming could we use?
Because perhaps with "proper" naming it would make more sense to put it on the container?

- `selectionChanged(array)` in one of the next breaking releases? and for now `select(array)`?
- `selectItem(item)` / `itemSelected(item)`
- `deselectItem(item)` / `itemDeselected(item)`

Reply:
Yes I would probably go with `itemSelected` and `itemDeselected` as this won't introduce a breaking change by renaming `select`. But I always tried to avoid past tense because I always thought it's rather a code smell. Yet I looked at a few popular libraries including the Angular CDK and Angular Material and they use past tense all the time, though not consistently.

Reply:
I think since you programatically select and deselect items on the container it would be a good idea to move those outputs to the container too 🤔 Although semantically I would prefer them on the directive because after all those are the items being selected. But this reasoning would also apply for the `select` and `deselect` API on the container. But this would be tricky to move to the directive because I don't want to tightly couple the item to the container by injecting the container into the directive. Another way would be to use a service but that introduce a lot of code and extra overhead which makes the library less lightweight.

Reply:
Another thought: what about keeping it on the directive, but only emitting a value on a user interaction, eg if you use `container.select` it won't a value.

I'm leaning more towards leaving it as is, on the directive. And indeed I think adding an extra service is a bit over kill.

Reply:
So Kwinten says it would make more sense for him to move those outputs to the container, but it's more of a gut feeling. Maybe you're right. Maybe it makes more sense on the container just to say consistent here.

Reply:
Ugh, this is really a hard one 😅 , then let's add it to the container?

Reply:
Yep maybe makes sense.

Reply:
Although... 😆My gut feeling says it makes more sense on the directive, but you select on the container.

Reply:
Then I got your back, and we should follow your gut feeling :)